scandium-core/src/main/java/org/eclipse/californium/scandium/dtls/EcdhPskServerKeyExchange.java

```diff
@@ -0,0 +1,247 @@
+	private ECPublicKey publicKey = null;
+
+	private ECPoint point = null;
+	private byte[] pointEncoded = null;
+
+	private final int curveId;
+	
+	// TODO right now only named curve is supported
+	private int curveType = NAMED_CURVE;
+	
+	/**
+	 * Creates a new key exchange message with ServerDHParams.
+	 * See <a href="http://tools.ietf.org/html/rfc4279#section-3">RFC 4279</a>:
+	 * 
+	 * @param ecdhe - {@code ECDHECryptography}
+	 * @param clientRandom - nonce
+	 * @param serverRandom - nonce
+	 * @param namedCurveId - ec curve used 
+	 * @param peerAddress
+	 * @throws GeneralSecurityException
+	 * @throws NullPointerException if any of arguments ecdhe, 
+	 * clientRandom, serverRandom are {@code null}
+	 */
+	public EcdhPskServerKeyExchange(ECDHECryptography ecdhe, Random clientRandom, Random serverRandom, 
+			int namedCurveId, InetSocketAddress peerAddress) throws GeneralSecurityException {	
+		this(namedCurveId, peerAddress);
+		if (ecdhe == null) {
+			throw new NullPointerException("ECDHECryptography class object cannot be null");
+		}
+		if (clientRandom == null || serverRandom == null) {
+			throw new NullPointerException("nonce cannot be null");
+		}
+		publicKey = ecdhe.getPublicKey();
+		ECParameterSpec parameters = publicKey.getParams();
+		point = publicKey.getW();
+		pointEncoded = ECDHECryptography.encodePoint(point, parameters.getCurve());
+	}
+	
+	private EcdhPskServerKeyExchange(int curveId, byte[] pointEncoded, InetSocketAddress peerAddress) throws HandshakeException {		
+		this(curveId, peerAddress);
+		if (pointEncoded == null) {
+			throw new NullPointerException("ephemeral public key cannot be null");
+		}
+		this.pointEncoded = Arrays.copyOf(pointEncoded, pointEncoded.length);
+		// re-create public key from params
+		SupportedGroup group = SupportedGroup.fromId(curveId);
+		if (group == null || !group.isUsable()) {
+			throw new HandshakeException(
+				String.format("Server used unsupported elliptic curve (%d) for ECDH", curveId),
+				new AlertMessage(AlertLevel.FATAL, AlertDescription.HANDSHAKE_FAILURE, peerAddress));
+		} else {
+			try {
+				point = ECDHECryptography.decodePoint(pointEncoded, group.getEcParams().getCurve());
+				KeyFactory keyFactory = KeyFactory.getInstance(KEYPAIR_GENERATOR_INSTANCE);
+				publicKey = (ECPublicKey) keyFactory.generatePublic(new ECPublicKeySpec(point, group.getEcParams()));
+			} catch (GeneralSecurityException e) {
+				LOGGER.debug("Cannot re-create server's public key from params", e);
+				throw new HandshakeException(
+					String.format("Cannot re-create server's public key from params: %s", e.getMessage()),
+					new AlertMessage(AlertLevel.FATAL, AlertDescription.INTERNAL_ERROR, peerAddress));
+			}
+		}
+	}
+
+	private EcdhPskServerKeyExchange(int namedCurveId, InetSocketAddress peerAddress) {
+		super(peerAddress);
+		this.curveId = namedCurveId;
+	}
+
+	// TODO this is called 4 times for Flight 4
+	@Override
+	public byte[] fragmentToByteArray() {
```

Comment:
the byte array produced by this method does not comply with https://tools.ietf.org/html/rfc5489#section-2 FMPOV since it does not contain the (potentially empty) identity hint. Even if there is no identity hint, you need to at least write the length of the hint (i.e. 0 bytes) to the array. Compare with PSKServerKeyExchange

Reply:
implemented